nabs/suspenders.py

```diff
@@ -0,0 +1,179 @@
+
+
+class SuspendPreprocessor:
+    """
+    Base preprocessor that suspends on specific commands based on a signal.
+
+    Parameters
+    ----------
+    signal: ``Signal``
+        The signal to subscribe to, whose value determines when to suspend.
+
+    commands: ``list of str``, optional
+        The commands to suspend on. If omitted, we'll suspend on all commands.
+
+    sleep: ``int`` or ``float``, optional
+        The amount of time to wait after `should_resume` returns ``True``
+        before ending the suspension. If `should_suspend` return ``True`` at
+        any time during this wait period, we will cancel the resumption and
+        return to the suspended state.
+    """
+    def __init__(self, signal, *, commands=None, sleep=0,
+                 pre_plan=null, post_plan=null, follow_plan=null):
+        self._sig = signal
+        self._cmd = commands
+        self._sleep = sleep
+        self._suspend_active = False
+        self._resume_ts = None
+        self._suspend_ev = Event()
+        self._ok_future = asyncio.Future()
+        self._ok_future.set_result('ok')
+        self._rlock = RLock()
+        self._subid = None
+
+    def should_suspend(self, value):
+        """
+        Returns ``True`` if we should suspend.
+
+        Parameters
+        ----------
+        value: signal value
+            The value reported by a signal callback.
+        """
+        raise NotImplementedError()
+
+    def should_resume(self, value):
+        """
+        Returns ``True`` if we should resume.
+
+        Parameters
+        ----------
+        value: signal value
+            The value reported by a signal callback.
+        """
+        return not self.should_suspend(value)
+
+    def _update(self, *, value, **kwargs):
+        """
+        Update routine for when the signal's value changes.
+
+        If we're running normally but we should_suspend, we'll trigger the
+        suspend state. If suspended but we should_resume, we'll start a timer
+        of length sleep to clear the suspend state. This will be interrupted if
+        another value change leads us to should_suspend.
+        """
+        with self._rlock:
+            if self._suspend_ev.is_set():
+                if self.should_resume(value):
+                    self._suspend_ev.clear()
+                    self._run_release_thread()
+            else:
+                if self.should_suspend(value):
+                    logger.info('Suspending due to bad %s value=%s',
+                                self._sig.name, value)
+                    loop = asyncio.get_event_loop()
+                    self._ok_future = loop.create_future()
+                    self._suspend_ev.set()
+
+    def _run_release_thread(self):
+        logger.info('%s suspension is over, waiting for %ss then resuming.',
+                    self._sig.name, self._sleep)
+        loop = asyncio.get_event_loop()
+        loop.run_in_executor(None, self._release_thread)
```

Comment:
Should this all clear any prior release threads? Imagining the following situation.

1. Suspender fires
2. Suspender returns back to operating region
2. Create release thread that waits for 5 seconds
3. Supsender fires
4. Suspender returns back to operation region, spawns a new 5 second wait
3. Release thread from first suspension resolves in less that 5 seconds after second suspension

Reply:
Sounds good. Thanks for the clarification.